features/323-login-id/api.md

```diff
@@ -0,0 +1,170 @@
++ async signupWithEmail(email: string, password: string, data?: object, realm?: string): Promise<User>;
+
+- async login(loginID: { [key: string]: string }, password: string): Promise<User>;
+- async loginWithUsername(username: string, password: string): Promise<User>;
+- async loginWithEmail(email: string, password: string): Promise<User>;
++ async login(loginID: { [key: string]: string } | string, password: string, realm?: string): Promise<User>;
++ async loginWithUsername(username: string, password: string, realm?: string): Promise<User>;
++ async loginWithEmail(email: string, password: string, realm?: string): Promise<User>;
+
++ async requestVerification(loginIDKey: string, loginID: string): Promise<void>;
++ async createLoginID(loginIDKey: string, loginID: string, realm?: string): Promise<void>;
++ async deleteLoginID(loginIDKey: string, loginID: string, realm?: string): Promise<void>;
+
+  async whoami(): Promise<User>;
+```
+
+### Example user object
+```json
+{
+    "user_id": "5bf1e4d2-e1c4-4517-93c7-7dae89261da6",
+    "metadata": {
+        "email": "test@example.com"
+    },
+    "last_seen_at": "...",
+    "created_at": "...",
+    "created_by": "...",
+    "updated_at": "...",
+    "updated_by": "...",
+    "verified": true,
+    "verify_info": {
+        "test@example.com": true
+    },
+    "access_token": "..."
+}
+```
+
+## Use Case Examples
+
+Unless specified, following examples assume using default configuration.
+
+### Signup with single login ID
+```typescript
+const user1 = await signupWithEmail("test@example.com", "12345678");
+// user1.metadata is { "email": "test@example.com" }
+const user2 = await signup({ phone: "+85299999999" }, "12345678");
+// user2.metadata is { "phone": "+85299999999" }
+
+// following should fail:
+await signup({ fingerprint: "ZmluZ2VycHJpbnQ=" }, "12345678"); // not white-listed
+```
+
+### Signup with realms
+Assume `teacher` and `student` are whitelisted realms.
+```typescript
+await signupWithEmail("test@example.com", "12345678", undefined, "teacher");
+await createLoginID("email", "test@example.com", "student");
+// then:
+
+// - following should succeed:
+await loginWithEmail("test@example.com", "12345678", "teacher");
+await loginWithEmail("test@example.com", "12345678", "student");
+
+// - following should fail:
+await loginWithEmail("test@example.com", "12345678"); // realm not matched
+await signupWithEmail("test@example.com", "12345678", undefined, "admin"); // cannot create login ID (use createLoginID instead)
+await signupWithEmail("test@example.com", "12345678"); // cannot reuse login ID for new user
+```
+
+### Signup with multiple login ID
+Assume `email` and `username` are white-listed login ID keys,
+with exactly one username allowed.
+
+```typescript
+const user = await signup([
+    {"email": "test@oursky.com"},
+    {"email": "test@example.com"},
+    {"username": "test"}
+], "12345678");
+// user.metadata is { "email": "test@oursky.com", "username": "test" }
```

Comment:
I think in existing design, loginID are not automatically added to `user.metadata`, is it changed or my understanding was incorrect?

Reply:
Yes, it is changed to avoid the concept of persistent 'primary login ID'; please refer to stage 3 for detail.

Reply:
Okay. I consider stage1-3.md are about implementation details, not part of the spec. For the sack of convinence for reference in future, can we centralize all spec (user's interface related) in this markdown?

I'm a bit skeptical (or worried) about auto-populating metadata with login ID actually... we can try, not sure how users would react.

Reply:
But we need to decide if we still want to auto-populating metadata before public launch. @kiootic can you put up an issue about it and mark it in the launch milestone? Something like: "Decide if Skygear Auth should autopopulate metadata with loginIDs"

Reply:
ref #331